### PR TITLE
Make each bank info form's rendered markup unique

### LINKED ISSUE #19594: 508-defect-1 ❗Launchblocker [AXE-CORE, COGNITION, SEMANTIC MARKUP]: Form fields and their unique IDs MUST NOT be duplicated on the page
# [508-defect-1 :exclamation: Launchblocker](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/platform/accessibility/guidance/defect-severity-rubric.md#508-defect-1)

## Feedback framework

- **❗️ Must** for if the feedback must be applied 
- **⚠️ Should** if the feedback is best practice 
- **✔️ Consider** for suggestions/enhancements 

## Definition of done

1. Review and acknowledge feedback.
1. Fix and/or document decisions made.
1. Accessibility specialist will close ticket after reviewing documented decisions / validating fix.

## Point of Contact

**VFS Point of Contact:** _Josh_

## Axe Violations

**[IDs used in ARIA and labels must be unique - Critical](https://dequeuniversity.com/rules/axe/3.5/duplicate-id-aria?application=coconut)**
Duplicate ARIA IDs are common validation errors that may break the accessibility of labels, e.g., form fields, table header cells. Unique ID's differentiate each element from another and prevent invalid markup, wherein only the first instance gets acted upon by client-side scripting, or where assistive technologies typically only reference the first one accurately.

**[Form fields do not have duplicate labels - Moderate](https://dequeuniversity.com/rules/axe/3.5/form-field-multiple-labels?application=coconut)**
Assigning multiple labels to the same form field can cause problems for some combinations of screen readers and browsers, and the results are inconsistent from one combination to the next. Some combinations will read the first label. Some will read the last label. Others will read both labels.

**[ID attribute values must be unique - Minor](https://dequeuniversity.com/rules/axe/3.5/duplicate-id?application=coconut)**
The ID attribute uniquely identifies elements on a page. It does not make sense to duplicate an ID.

Duplicate ID's can break the accessibility of labels for forms, table header cells, etc., by the second instance being skipped by screen readers, or by client-side scripts. They are common markup validation errors that can eliminate possible sources of accessibility problems, when not breaking accessibility.

Those experienced with client-side scripting know that when you re-use an id, typically the only one that gets acted upon by the scripting is the first instance of the use of that ID. Similarly, assistive technologies may, when referencing an id, only reference the first one accurately.

## Details

It seems like fields were duplicated for all the edit labels and fields within disability compensation and education benefits. This resulted in a number of axe violations that must be resolved before launch.

## Acceptance Criteria

- [ ] All axe violations listed are resolved

## Environment

* Operating System: any
* Browser: all
* Screenreading device: any
* Server destination: staging

## Steps to Recreate

1. Enter `https://staging.va.gov/profile/direct-deposit` in browser
2. Open the edit states for both sections
3. Run axe-coconut
4. Confirm "IDs used..." error is flagged for each duplicated field (3)
5. Confirm "id attributes..." error is flagged for labels and error (4)
6. Confirm "Form fields..." error is flagged for all form fields (6)

## Proposed Solution

**[IDs used in ARIA and labels must be unique - Critical](https://dequeuniversity.com/rules/axe/3.5/duplicate-id-aria?application=coconut)**
Rename any duplicate ARIA ID values. Duplicate ARIA IDs are common validation errors that may break the accessibility of labels, e.g., form fields, table header cells. To fix the problem, change an ID value if it is used more than once to be sure each is unique. Unique ID's differentiate each element from another and prevent invalid markup, wherein only the first instance gets acted upon by client-side scripting, or where assistive technologies typically only reference the first one accurately.

**[Form fields do not have duplicate labels - Moderate](https://dequeuniversity.com/rules/axe/3.5/form-field-multiple-labels?application=coconut)**
Ensure that only one label is assigned to a form field.

**[ID attribute values must be unique - Minor](https://dequeuniversity.com/rules/axe/3.5/duplicate-id?application=coconut)**
Rename any duplicate ID attributes values.

Duplicate IDs are common validation errors that may break the accessibility of labels, e.g., form fields, table header cells.

To fix the problem, change an ID value if it is used more than once to be sure each is unique. Unique ID's differentiate each element from another and prevent invalid markup, wherein only the first instance gets acted upon by client-side scripting, or where assistive technologies typically only reference the first one accurately.

Proper markup eliminates at least one possible source of accessibility problems. WCAG 1.0 used to have a provision which explicitly required the use of valid markup. WCAG 2.0 no longer requires valid markup. We mention it here because valid markup is a quick path toward ensuring accessibility. There are hundreds of ways a person can create invalid markup. Most of these ways are usually inconsequential for accessibility such as un-encoded ampersands, for instance. Others, however, are significant and may cause issues relating to how assistive technology interacts with the page and renders the page to the end user, such as subsequent instances of element attribute ID's being ignored.

Submit markup code to the W3C validator at http://validator.w3.org to the validity of HTML markup in order to quickly identify reused attribute ID values.

## WCAG or Vendor Guidance

* [IDs used in ARIA and labels must be unique - Critical](https://dequeuniversity.com/rules/axe/3.5/duplicate-id-aria?application=coconut)
* [Form fields do not have duplicate labels - Moderate](https://dequeuniversity.com/rules/axe/3.5/form-field-multiple-labels?application=coconut)
* [ID attribute values must be unique - Minor](https://dequeuniversity.com/rules/axe/3.5/duplicate-id?application=coconut)

## Screenshots or Trace Logs

Screen Shot 2021-02-08 at 11 00 17 AM
Screen Shot 2021-02-08 at 11 09 31 AM

### LINKED ISSUE #19790: 508-defect-1 ❗ Launchblocker [AXE-CORE]: Error id attribute value must be unique
# [508-defect-1 ❗ Launchblocker](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/platform/accessibility/guidance/defect-severity-rubric.md#508-defect-2)

## Feedback framework

- **❗️ Must** for if the feedback must be applied 
- **⚠️ Should** if the feedback is best practice 
- **✔️ Consider** for suggestions/enhancements 

## Definition of done

1. Review and acknowledge feedback.
1. Fix and/or document decisions made.
1. Accessibility specialist will close ticket after reviewing documented decisions / validating fix.

## Point of Contact

**VFS Point of Contact:** _Josh_

## Axe Violation

**[ID attribute values must be unique - Minor](https://dequeuniversity.com/rules/axe/3.5/duplicate-id?application=coconut)**
The ID attribute uniquely identifies elements on a page. It does not make sense to duplicate an ID.

Duplicate ID's can break the accessibility of labels for forms, table header cells, etc., by the second instance being skipped by screen readers, or by client-side scripts. They are common markup validation errors that can eliminate possible sources of accessibility problems, when not breaking accessibility.

Those experienced with client-side scripting know that when you re-use an id, typically the only one that gets acted upon by the scripting is the first instance of the use of that ID. Similarly, assistive technologies may, when referencing an id, only reference the first one accurately.

## Acceptance Criteria

- [ ] Axe no longer flags a violation for the error message

## Steps to Recreate

1. Recreate the error by editing and saving the first account
2. Run axe scan
3. Confirm "ID attribute values must be unique" is flagged for the error message

## Proposed Solution

Use a unique id attribute value

## Screenshots or Trace Logs

Screen Shot 2021-02-10 at 3 41 29 PM
Screen Shot 2021-02-10 at 3 49 16 PM

### PR DESCRIPTION
**Pull request by erikphansen**
## Description
Fixes https://github.com/department-of-veterans-affairs/va.gov-team/issues/19790 and fixes https://github.com/department-of-veterans-affairs/va.gov-team/issues/19594

Makes is possible to generate unique edit bank info forms, with unique IDs.

## Testing done
Local + Cypress tests

## Screenshots
N/A

## Acceptance criteria
- [ ]

## Definition of done
- [ ] Events are logged appropriately
- [ ] Documentation has been updated, if applicable
- [x] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)
- [x] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

### CODE REVIEW

File: src/applications/personalization/profile/tests/e2e/direct-deposit/v1-update-flow.cypress.spec.js [hunk 1]
```diff
@@ -0,0 +1,98 @@
+import disableFTUXModals from '~/platform/user/tests/disableFTUXModals';
```
Comment (erikphansen) [inline src/applications/personalization/profile/tests/e2e/direct-deposit/v1-update-flow.cypress.spec.js:1]: This entire file is mostly a copy-paste job from the existing `update-flow.cypress` test. It's just running against the old/existing version of Direct Deposit that only supports Direct Deposit for Comp and Pen.

File: src/applications/personalization/profile/tests/e2e/direct-deposit/v1-update-flow.cypress.spec.js [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+import mockDD4EDUEnrolled from '@@profile/tests/fixtures/dd4edu/dd4edu-enrolled.json';
+
+// this entire test file can be removed when DD4EDU is on prod to 100% of users
+const dd4eduDisabled = {
+  data: {
+    type: 'feature_toggles',
+    features: [
+      {
+        name: 'ch33_dd_profile',
+        value: false,
```
Comment (erikphansen) [inline src/applications/personalization/profile/tests/e2e/direct-deposit/v1-update-flow.cypress.spec.js:16]: Making sure this feature flag is turned off so we got the old version of Direct Deposit

File: src/applications/personalization/profile/tests/e2e/direct-deposit/v1-update-flow.cypress.spec.js (hunk 1 above)
Comment (erikphansen) [reply src/applications/personalization/profile/tests/e2e/direct-deposit/v1-update-flow.cypress.spec.js:1]: I had to add this test to make sure I wasn't breaking things that were in production since I was making changes that turned out to be non-trivial and a little bit scary :-D

File: src/applications/personalization/profile/components/direct-deposit/BankInfoCNP.jsx [hunk 3]
```diff
@@ -51,9 +52,31 @@ export const BankInfoCNP = ({
   const isSavingBankInfo = directDepositUiState.isSaving;
   const saveError = directDepositUiState.responseError;
 
-  const { accountNumber, accountType, routingNumber } = formData;
-  const isEmptyForm = !accountNumber && !accountType && !routingNumber;
+  const bankInfoUpdatedAlertSettings = {
```
Comment (erikphansen) [inline src/applications/personalization/profile/components/direct-deposit/BankInfoCNP.jsx:55]: I had to add this stuff to make the Cypress test work nicely. It's how the new Direct Deposit is set up.

File: src/applications/personalization/profile/components/direct-deposit/BankInfoForm.jsx [hunk 4]
```diff
@@ -6,93 +6,117 @@ import LoadingButton from 'platform/site-wide/loading-button/LoadingButton';
...
-      enum: Object.values(ACCOUNT_TYPES_OPTIONS),
+export function makeFormProperties(prefix) {
+  return {
+    routingNumber: `${prefix}RoutingNumber`,
+    accountNumber: `${prefix}AccountNumber`,
+    accountType: `${prefix}AccountType`,
+  };
+}
+
+function makeSchemas(prefix) {
```
Comment (erikphansen) [inline src/applications/personalization/profile/components/direct-deposit/BankInfoForm.jsx:17]: This method, making it possible to make different versions of the form schemas, is the crux of this PR.

File: src/applications/personalization/profile/tests/components/direct-deposit/BankInfoForm.unit.spec.jsx [hunk 5]
```diff
@@ -36,10 +33,6 @@ describe('<BankInfoForm/>', () => {
   it('should render a SchemaForm', () => {
     expect(schemaForm.exists()).to.be.true;
   });
-  it("should set the SchemaForm's schema  and uiSchema correctly", () => {
```
Comment (erikphansen) [inline src/applications/personalization/profile/tests/components/direct-deposit/BankInfoForm.unit.spec.jsx:39]: This test was failing after my change and I am deleting it because I don't like it 😆 It's a prime example of testing implementation details. And it now fails because there is no longer a static `schema` or `uiSchema` defined alongside the `BankInfoForm`. But `BankInfoForm` still does exactly what it's supposed to do. A more useful test would be to make sure that the `BankInfoForm` renders the form inputs we expect it to and they have the correct IDs.

File: src/applications/personalization/profile/components/direct-deposit/BankInfoForm.jsx [hunk 6]
```diff
@@ -6,93 +6,117 @@ import LoadingButton from 'platform/site-wide/loading-button/LoadingButton';
...
-    onSubmit={formSubmit}
-  >
-    <LoadingButton
-      type="submit"
-      className="usa-button-primary vads-u-margin-top--0 vads-u-width--full small-screen:vads-u-width--auto"
-      isLoading={isSaving}
-    >
-      Update
-    </LoadingButton>
+}) => {
```
Comment (erikphansen) [inline src/applications/personalization/profile/components/direct-deposit/BankInfoForm.jsx:77]: most of this is diff noise due to changing some indentation.

File: src/applications/personalization/profile/tests/e2e/direct-deposit/update-flow.cypress.spec.js [hunk 7]
```diff
@@ -19,11 +19,17 @@ const dd4eduEnabled = {
   },
 };
 
-function fillInBankInfoForm() {
+function fillInBankInfoForm(id) {
```
Comment (erikphansen) [inline src/applications/personalization/profile/tests/e2e/direct-deposit/update-flow.cypress.spec.js:22]: I added a test where _both_ edit bank info forms are visible at the same time so I needed a way to target _which_ bank info form is being filled in.

File: src/applications/personalization/profile/tests/e2e/direct-deposit/update-flow.cypress.spec.js [hunk 8]
```diff
@@ -36,8 +42,15 @@ function exitBankInfoForm() {
   cy.findByRole('button', { name: /cancel/i }).click();
 }
 
-function saveNewBankInfo() {
-  cy.findByRole('button', { name: /update/i }).click();
+function saveNewBankInfo(id) {
```
Comment (erikphansen) [inline src/applications/personalization/profile/tests/e2e/direct-deposit/update-flow.cypress.spec.js:45]: Same deal here. We need a way to determine which form's UPDATE button we are clicking.

File: src/applications/personalization/profile/tests/e2e/direct-deposit/update-flow.cypress.spec.js [hunk 9]
```diff
@@ -117,4 +133,34 @@ describe('Direct Deposit', () => {
       cy.axeCheck();
     });
   });
+  describe('when editing both at the same time and they both fail to update', () => {
```
Comment (erikphansen) [inline src/applications/personalization/profile/tests/e2e/direct-deposit/update-flow.cypress.spec.js:136]: This new test makes sure we fixed the a11y issues this PR was attempting to fix. Namely: with both edit bank info forms visible, does it pass an aXe scan? And also that when both form's error alerts are shown, does it pass an aXe scan?

File: src/applications/personalization/profile/components/direct-deposit/BankInfoCNPv2.jsx [hunk 10]
```diff
@@ -217,7 +230,7 @@ export const BankInfoCNP = ({
   // account information
   const editingBankInfoContent = (
     <>
-      <div id="errors" role="alert" aria-atomic="true">
+      <div id="cnp-bank-save-errors" role="alert" aria-atomic="true">
```
Comment (sanlouise) [inline src/applications/personalization/profile/components/direct-deposit/BankInfoCNPv2.jsx:233]: Not part of this PR, but TIL about `aria-atomic`!

**Review by sanlouise — APPROVED** [review]
